### PR TITLE
Type Lambdas Spec

### PR DESCRIPTION
**Pull request by odersky**

### CODE REVIEW

File: docs/docs/reference/type-lambdas-spec.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+Only the upper bound `H` can be F-bounded, i.e. `X` can appear in it.
+
+A variance annotation on a parameter indicates a subtyping relationship on type instances. For instance, given
+```
+type TL1 = [+A] => F[A]
+type TL2 = [-A] => F[A]
+```
+and two types `S <: T`, we have
+```
+TL1[S] <: TL1[T
```
Comment (smarter) [inline docs/docs/reference/type-lambdas-spec.md]: missing ]

**Comment by odersky** [comment]
Any comments? If not let's get this in. We have lots of other parts to work on as well.

File: docs/docs/reference/type-lambdas-spec.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+## Subtyping Rules
+
+Assume two type lambdas
+```scala
+type TL1  =  [v1 X >: L1 <: U1] => R1
+type TL2  =  [v2 X >: L2 <: U2] => R2
+```
+where `v1` and `v2` are optional variance annotations: `+`, `-`, or absent.
+Then `TL1 <: TL2`, if the type interval `L2..U2` is contained in the type interval `L1..U1` (i.e.
+`L1 <: L2` and `U2 <: U1`), and either `v2` is absent or `v1 = v2`.
```
Comment (sjrd) [inline docs/docs/reference/type-lambdas-spec.md]: The subtyping rule says nothing about `R1` and `R2`. That looks suspicious to me. Shouldn't `R1 <: R2`?

**Comment by milessabin** [comment]
I second @smarter's comment.

File: docs/docs/reference/type-lambdas-spec.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+A parameterized abstract type
+```scala
+type T[X] >: L <: U
+```
+is regarded as shorthand for an unparameterized abstract type with type lambdas as bounds.
+```scala
+type T >: ([X] => L) <: ([X] => U)
+```
+However, if `L` is `Nothing` it is not parameterized, since `Nothing` is treated as a bottom type for all kinds. For instance,
```
[the remark this answers is not in the document]
Comment (odersky) [reply docs/docs/reference/type-lambdas-spec.md]: It's a design choice. Whether it's necessary or not can be determined only by trying to change it and observing what breaks 😉

File: docs/docs/reference/type-lambdas-spec.md [hunk 4]
```diff
@@ -0,0 +1,111 @@
+---
+layout: doc-page
+title: "Type Lambdas - More Details"
+---
```
Comment (nicolasstucki) [inline docs/docs/reference/type-lambdas-spec.md]: This document should be added to `docs/sidebar.yml`.

**Comment by odersky** [comment]
@nicolasstucki No, spec pages should not be part of sidebar.yml. They should be referred to from the main page of the topic.

**Pull request merged**